compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -137,32 +137,162 @@
                   node);                                                           \
       }
 
-/**
- * Binary identity or zero operation
- *
- * If the second child is a constant that represents an identity operation,
- * replace this node with the first child.
- *
- * If the second child is a constant that represents a zero operation,
- * replace this node with the second child.
- */
-#define BINARY_IDENTITY_OR_ZERO_OP(ConstType,Type,NullValue,ZeroValue)\
-   if(secondChild->getOpCode().isLoadConst())\
-      {\
-      ConstType value = secondChild->get##Type();\
-      if(value == NullValue)\
-         return s->replaceNodeWithChild(node, firstChild, s->_curTree, block);\
-      if(value == ZeroValue)\
-         {\
-         if(performTransformation(s->comp(), "%sFound op with iconst in node [" POINTER_PRINTF_FORMAT "]\n", s->optDetailString(),node))\
-            {\
-            s->anchorChildren(node, s->_curTree);\
-            s->prepareToReplaceNode(node, secondChild->getOpCodeValue());\
-            node->set##Type(value);\
-            return node;\
-            }\
-         }\
+int8_t getNodeByteValue(TR::Node * node)
+   {
+   return node->getByte();
+   }
+
+int16_t getNodeShortValue(TR::Node * node)
+   {
+   return node->getShortInt();
+   }
+
+int32_t getNodeIntValue(TR::Node * node)
+   {
+   return node->getInt();
+   }
+
+int64_t getNodeLongValue(TR::Node * node)
+   {
+   return node->getLongInt();
+   }
+
+uint32_t getNodeFloatBits(TR::Node * node)
+   {
+   return node->getFloatBits();
+   }
+
+uint64_t getNodeDoubleBits(TR::Node * node)
+   {
+   return node->getDoubleBits();
+   }
+
+void setNodeByteValue(TR::Node * node, int8_t value)
+   {
+   node->setByte(value);
+   }
+
+void setNodeShortValue(TR::Node * node, int16_t value)
+   {
+   node->setShortInt(value);
+   }
+
+void setNodeIntValue(TR::Node * node, int32_t value)
+   {
+   node->setInt(value);
+   }
+
+void setNodeLongValue(TR::Node * node, int64_t value)
+   {
+   node->setLongInt(value);
+   }
+
+void setNodeFloatBits(TR::Node * node, uint32_t bits)
+   {
+    node->setFloatBits(bits);
+   }
+
+void setNodeDoubleBits(TR::Node * node, uint64_t bits)
+   {
+   node->setDoubleBits(bits);
+   }
+
+inline bool isNodeLoadConst(TR::Node * node)
+   {
+   return node != NULL && node->getOpCode().isLoadConst();
+   }
+
+template <typename T>
+struct BinaryOpSimplifier
+   {
+   TR::Simplifier * s;
+   T (* getNodeValue)(TR::Node *);
+   void (* setNodeValue)(TR::Node *, T);
+
+   /**
+   * Binary identity operation
+   *
+   * If the second child is a constant that represents an identity operation,
+   * replace this node with the first child.
+   */
+   inline TR::Node * tryAndSimplifyIdentityOp(TR::Node * node, T nullValue)
```

Comment:
I think it makes more sense to the pass the parent node in (`node`) since that's the binary op we are trying to simplify. It has better encapsulation since you can't accidentally pass in the wrong combination of `node, firstChild, secondChild`, the code will always appropriately get the first and second child from the parent.